book/src/how/r1cs_constraints.md

```diff
@@ -2,49 +2,72 @@
-* The lookup output. 
+$Az(r_x) = \sum_y A(r_x, y) * z(y)$
 
-### Circuit and instruction flags: 
-* There are nine circuit flags used to guide the constraints and are dependent only on the opcode of the instruction. These are thus stored as part of the preprocessed bytecode in Jolt. 
-    1. `operand_x_flag`: 0 if the first operand is the value in `rs1` or the `PC`. 
-    2. `operand_y_flag`: 0 if the second operand is the value in `rs2` or the `imm`. 
-    3. `is_load_instr`
-    4. `is_store_instr`
-    5. `is_jump_instr`
-    6. `is_branch_instr`
-    7. `if_update_rd_with_lookup_output`: 1 if the lookup output is to be stored in `rd` at the end of the step. 
-    8. `sign_imm_flag`: used in load/store and branch instructions where the instruction is added as constraints. 
-    9. `is_concat`: indicates whether the instruction performs a concat-type lookup. 
-* Instruction flags: these are the unary bits used to indicate instruction is executed at a given step. There are as many per step as the number of unique instruction lookup tables in Jolt, which is 19. 
+$Bz(r_x) = \sum_y B(r_x, y) * z(y)$
 
-#### Constraint system 
+$Cz(r_x) = \sum_y C(r_x, y) * z(y)$
 
-The constraints for a CPU step are detailed in the `get_jolt_matrices()` function in [`constraints.rs`](https://github.com/a16z/jolt/blob/main/jolt-core/src/r1cs/constraints.rs). 
+We could run a sumcheck for each of these but it's more efficient from both a prover and verifier perspective to combine into a single sumcheck by pulling a random variable $r_c$ from the Fiat-Shamir transcript and combining the linear terms.
+
+$ABCz(r_x) = \widetilde{Az}(r_x) + r_c * \widetilde{Bz}(r_x)  + r_c^2 * \widetilde{Cz}(r_x) = \sum_y{[ A(r_x, y) + r_c * B(r_x, y) + r_c^2 * C(r_x, y] ) * \widetilde{Z}(y)}$
+
+Outputs a claimed evaluation of $A(r_x, r_y), B(r_x, r_y), C(r_x, r_y), Z(r_y)$, specifically in the form $A(r_x,r_y) + r_c * B(r_x, r_y) + r_c^2 * C(r_x, r_y)$.
+
+The verifier does have a representation of these is able to evaluate without assistance.
+
+## Uniform Spartan
+Jolt's Spartan instance is highly patterned. We run the same R1CS for each cycle of the RISC-V program. We can leverage this property to increase prover and verifier efficiency.
+
+Think of a simple constraint system with the following inputs: $Q, R, S$. Each of these inputs is repeated $nc = num\_cycles$ times. 
+
+The witness vector, $Z$ now looks like $[Q_0, ... Q_{nc}, R_0, ... R_{nc}, S_{0}, ... S_{nc}, 1]$. The one in the final slot allows constants in the constraints.
+
+> [!NOTE]
+> Syntax: For bitvector $x=x_0, x_1, ... x_n$, $x_0$ represents the most-significant bit, and $x_n$ represents the least significant bit. By default sumchecks will bind from the "top", where they bind the MSB first. For some sumchecks, notably sparse ones we'll bind from the "bottom" for algorithmic efficiency reasons. 
```

Comment:
I'm hoping to switch all sumchecks to bind from the bottom so this note may soon become out-of-date